billy-portugal/src/main/java/com/premiumminds/billy/portugal/services/export/webservice/AuthenticationHandler.java

```diff
@@ -0,0 +1,185 @@
+        SOAPFactory soapFactory = SOAPFactory.newInstance();
+
+        SOAPElement wsSecHeaderElm = soapFactory.createElement("Security", AUTH_PREFIX, AUTH_NS);
+        SOAPElement userNameTokenElm = soapFactory.createElement("UsernameToken", AUTH_PREFIX, AUTH_NS);
+        SOAPElement userNameElm = soapFactory.createElement("Username", AUTH_PREFIX, AUTH_NS);
+        userNameElm.addTextNode(username);
+        SOAPElement passwdElm = soapFactory.createElement("Password", AUTH_PREFIX, AUTH_NS);
+        passwdElm.addTextNode(b64EncryptedPassword);
+        SOAPElement nonceElm = soapFactory.createElement("Nonce", AUTH_PREFIX, AUTH_NS);
+        nonceElm.addTextNode(b64EncryptedSimetricKey);
+        SOAPElement createdElm = soapFactory.createElement("Created", AUTH_PREFIX, AUTH_NS);
+        createdElm.addTextNode(b64EncryptedTimestamp);
+
+        userNameTokenElm.addChildElement(userNameElm);
+        userNameTokenElm.addChildElement(passwdElm);
+        userNameTokenElm.addChildElement(nonceElm);
+        userNameTokenElm.addChildElement(createdElm);
+
+        wsSecHeaderElm.addChildElement(userNameTokenElm);
+
+        final SOAPHeader sh = envelope.getHeader();
+
+        sh.addChildElement(wsSecHeaderElm);
+    }
+
+    private byte[] cypherRequestKey(PublicKey publicKey, byte[] requestKey)
+            throws NoSuchAlgorithmException, NoSuchPaddingException, InvalidKeyException, IllegalBlockSizeException,
+            BadPaddingException {
+        Cipher cipher = Cipher.getInstance("RSA");
+        cipher.init(Cipher.ENCRYPT_MODE, publicKey);
+        return cipher.doFinal(requestKey);
+    }
+
+    public byte[] generateRequestKey() throws NoSuchAlgorithmException {
+        KeyGenerator generator = KeyGenerator.getInstance("AES");
+        generator.init(128);
+
+        Key keyToBeWrapped = generator.generateKey();
+
+        return keyToBeWrapped.getEncoded();
+    }
+
+    private byte[] cypherCredential(byte[] requestKey, String credential)
+            throws NoSuchAlgorithmException, NoSuchPaddingException, InvalidKeyException, IllegalBlockSizeException,
+            BadPaddingException {
+        Cipher cipher = Cipher.getInstance("AES/ECB/PKCS5Padding");
+
+        cipher.init(Cipher.ENCRYPT_MODE, new SecretKeySpec(requestKey, "AES"));
+
+        return cipher.doFinal(credential.getBytes(StandardCharsets.UTF_8));
+    }
+
+    private Key getPublicKeyFromKeystore()
+            throws KeyStoreException, NoSuchAlgorithmException, CertificateException, IOException {
+
+        KeyStore ks = KeyStore.getInstance("JKS");
+
+        try (InputStream fis = webserviceKeys.getKeystore().toURL().openStream()) {
+            ks.load(fis, webserviceKeys.getKeystorePassword().toCharArray());
+            return ks.getCertificate(webserviceKeys.getKeyAlias()).getPublicKey();
+        }
+    }
+
+    private static final SimpleDateFormat TIMESTAMP_FORMATER = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.S'Z'");
```

Comment:
Constants should be at the start of the file

Reply:
changed